### ISSUE TITLE
Generating Similar items for all items available in data using similarity matrix

### ISSUE THREAD

**Issue by himanshu-jainz** [benfred/implicit#364]
I generated the item weights using ALS functionality, now I need to create top 10 similar products  for all the products available in dataset.
I can use similar_items functionality and generate similar products one at a time in a loop but is there any smart functionality with the library that will be more fast?

**Comment by ita9naiwa** [comment]
precomputing similar items requires huge memory, so it can't be done simply. 

hopefully, there is good approximate neighbor search libraries such as Faiss, nmslib, and annoy.

implicit contains approximate ALS exploiting these approximate nearest neighbor libraries.
https://github.com/benfred/implicit/blob/11833275369f06c537f8d0acf5fc4b46465a8be9/implicit/approximate_als.py

**Comment by benfred** [comment]
The latest main branch of this repo has support for batch mode for the similar_items code, which lets you call similar_items for multiple items at once. This will run in multiple threads on the CPU - and can also take advantage of the GPU if you have one.  Both the CPU and GPU code will partition the itemids so that this fits into a reasonable amount of memory.

As an example of using this:

```python
# train a model
model = implicit.als.AlternatingLeastSquares()
model.fit(user_items)


# generate similar_items for all items in the model
itemids = np.arange(model.item_factors.shape[0])
ids, scores = model.similar_items(itemids)
```

Note that the API had to change to support this, so the return type of similar_items is different now - and the model.fit call takes a user_items matrix now instead of a item_users matrix.
